### ISSUE TITLE
Upgrading to the html-webpack-plugin version 4

### ISSUE THREAD

**Issue by jantimon** [waysact/webpack-subresource-integrity#80]
The html-webpack-plugin version 4 is getting closer to a release.  
Many information can be found in this pr: https://github.com/jantimon/html-webpack-plugin/pull/953

A static version property was added:
https://github.com/jantimon/html-webpack-plugin/blob/d65b37d2c588047e0d81a38f4645fcdb3ead0b9e/index.js#L915-L927

The event system was changed alot to work with the new features:  
<img src="https://raw.githubusercontent.com/jantimon/html-webpack-plugin/webpack-4/flow.png" />

Could you please give the new events a try and let me know if they work for your case?
https://github.com/jantimon/html-webpack-plugin/tree/webpack-4#events

#### `beforeAssetTagGeneration` hook

```
    AsyncSeriesWaterfallHook<{
      assets: {
        publicPath: string,
        js: Array<{string}>,
        css: Array<{string}>,
        favicon?: string | undefined,
        manifest?: string | undefined
      },
      outputName: string,
      plugin: HtmlWebpackPlugin
    }>
```

#### `alterAssetTags` hook

```
    AsyncSeriesWaterfallHook<{
      assetTags: {
        scripts: Array<HtmlTagObject>,
        styles: Array<HtmlTagObject>,
        meta: Array<HtmlTagObject>,
      },
      outputName: string,
      plugin: HtmlWebpackPlugin
    }>
```

#### `alterAssetTagGroups` hook

```
    AsyncSeriesWaterfallHook<{
      headTags: Array<HtmlTagObject | HtmlTagObject>,
      bodyTags: Array<HtmlTagObject | HtmlTagObject>,
      outputName: string,
      plugin: HtmlWebpackPlugin
    }>
```

#### `afterTemplateExecution` hook

```
    AsyncSeriesWaterfallHook<{
      html: string,
      headTags: Array<HtmlTagObject | HtmlTagObject>,
      bodyTags: Array<HtmlTagObject | HtmlTagObject>,
      outputName: string,
      plugin: HtmlWebpackPlugin,
    }>
```

#### `beforeEmit` hook

```
    AsyncSeriesWaterfallHook<{
      html: string,
      outputName: string,
      plugin: HtmlWebpackPlugin,
    }>
```

#### `afterEmit` hook

```
    AsyncSeriesWaterfallHook<{
      outputName: string,
      plugin: HtmlWebpackPlugin
    }>
```

**Comment by jscheid** [comment]
Hi @jantimon,

Thanks for the heads up. I've got this mostly working, but there is one issue:

In HWP 3.x I was using `beforeHtmlGeneration` to walk through all assets and add the special `Integrity` assets: https://github.com/waysact/webpack-subresource-integrity/blob/d7ef757b8da69be97e14dc997eb156af52841a51/index.js#L269-L280

If I hook that function into `beforeAssetTagGeneration` unchanged everything is fine, except for one small hitch: the following code to add the favicon asset is executed only after my hook into `beforeAssetTagGeneration` runs: https://github.com/jantimon/html-webpack-plugin/blob/d65b37d2c588047e0d81a38f4645fcdb3ead0b9e/index.js#L182-L190

This means that my hook doesn't "know" about that asset. While this is not necessarily a big deal as far as the favicon is concerned, I do worry that it might cause other assets (perhaps added by third parties) to be missed.

I'm going to try a few things to see if I can fiddle with the order of `tap` handlers to ensure that mine runs last. In the meantime, do you have any thoughts on this?

**Comment by jscheid** [comment]
@jantimon thanks for the heads up -- I had a different comment here before but I was chasing a red herring. All good. I'll release a version compatible with 4.0.0-alpha.2 soon.

**Comment by jscheid** [comment]
Preliminary compatibility released in 1.1.0-rc.5.

I'm planning to release 1.1.0 proper soon, perhaps timed to coincide with the HWP 4.0.0 release if that's not too far out?
